781 small issue compilation extract collections and groups

kkoehn opened the pull request:
Tackles most collection issues from #781 

- [x] add fast collection creation in task/show
- [x] prevent a user joining a collection multiple times (validation, error handling)
- [x] show count of users in collection/show (to preserve anonymity, since we have no collection admin)
- [x] rename leave button to delete, if user is last member of collection
- [x] `view_shared` didn't work at all
- [x] fix  tests
- [x] add tests

Additional tasks after review/discusstion:
- [x] add collection visibility
  - [x] default private => only members can see it
  - [x] optional public

Will be handled in different PR(s):
- [ ] show # of open invitations (since they can be used to get access at any time)
- [ ] consume invitations
- [ ] public collections can be saved by any user (but not edited) (follow)
- [ ] collection can be assigned to group (edit rights)

MrSerth commented:
> @MrSerth could you take a quick look at the changes (esp. the ui) to see if you have any suggestions?

Sure -- happy to do so. I think, the code mostly looks fine. Probably, one of the first things I would mention is to move all locales from the HTML, controller, ... to the respective locale files. Since we are planning to translate CodeHarbor to German in the future, we should already pay attention there.

Regarding the UI, I like the new option to create a collection. For the input / button, I would probably take a look at a [bootstrap input group with a button](https://getbootstrap.com/docs/5.3/forms/input-group/#button-addons) to make that look nice (and add some more padding to the last element of the list of existing collections). 

Other than that, I was wondering whether we should (in addition to the existing method) allow the assignment of a task to a collection through the tasks#edit route, similar to groups. There, a select2 (similar to labels) could be used, either showing existing groups or allowing to create a new one. This method (select2) could also be suitable for the tasks#show route, but I am not yet sure whether I prefer it.

When viewing collections, I was wondering whether there is some "user management": I am now seeing that I've shared my collection with X other users, but can I see whom I gave access or remove them?

kkoehn commented:
Reflection (more or less) of meeting on dec 20th

> Sure -- happy to do so. I think, the code mostly looks fine. Probably, one of the first things I would mention is to move all locales from the HTML, controller, ... to the respective locale files. Since we are planning to translate CodeHarbor to German in the future, we should already pay attention there.

done
> Regarding the UI, I like the new option to create a collection. For the input / button, I would probably take a look at a [bootstrap input group with a button](https://getbootstrap.com/docs/5.3/forms/input-group/#button-addons) to make that look nice (and add some more padding to the last element of the list of existing collections).

good idea, it looks a lot cleaner now
![image](https://github.com/openHPI/codeharbor/assets/24592181/c794695f-b143-4caf-9092-97e3ad9519a4)

> Other than that, I was wondering whether we should (in addition to the existing method) allow the assignment of a task to a collection through the tasks#edit route, similar to groups. There, a select2 (similar to labels) could be used, either showing existing groups or allowing to create a new one. This method (select2) could also be suitable for the tasks#show route, but I am not yet sure whether I prefer it.

I would prefer to keep the Collection addition in the show-view, since it should be easy and fast to add a Task to a Collection. I am not a fan of the group-handling either - it just feels confusing.

> When viewing collections, I was wondering whether there is some "user management": I am now seeing that I've shared my collection with X other users, but can I see whom I gave access or remove them?

If we add user management to Collection they basically become groups, therefore I would keep Collections a bit more simple and maybe a little anarchic. It would be nice if a member of the Collection could change its visibility between `private` and `public`, to enable other users to "follow" or "subscribe" a Collection without becoming an actual member.

MrSerth commented:
Thanks for tackling continuously working on this PR and addressing some of the points we discussed in late December. I think, the code and the functionality is fine. There are some optical things, but given upcoming rebrush, I probably wouldn't pay too much attention. The only exception might be the new Collections#show Route, where the three icons (within the orange box) all have other sizes, weights, and maybe the two lines of text are too close to each other (or the second line should be lighter...) I haven't tried anything, but it might be a bit too crowded: 

<img width="821" alt="Bildschirmfoto 2024-01-12 um 21 48 29" src="https://github.com/openHPI/codeharbor/assets/7300329/0d8910fb-a619-4056-a97e-f7d3adf514d1">

Other than checking the functionality (which seems to be fine), I haven't made a (code) review -- leaving that open for @Dome-GER.

kkoehn commented:
> Thanks for tackling continuously working on this PR and addressing some of the points we discussed in late December. I think, the code and the functionality is fine. There are some optical things, but given upcoming rebrush, I probably wouldn't pay too much attention. The only exception might be the new Collections#show Route, where the three icons (within the orange box) all have other sizes, weights, and maybe the two lines of text are too close to each other (or the second line should be lighter...) I haven't tried anything, but it might be a bit too crowded:
> <img alt="Bildschirmfoto 2024-01-12 um 21 48 29" width="821" src="https://private-user-images.githubusercontent.com/7300329/296361996-0d8910fb-a619-4056-a97e-f7d3adf514d1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDUxNDg2NTYsIm5iZiI6MTcwNTE0ODM1NiwicGF0aCI6Ii83MzAwMzI5LzI5NjM2MTk5Ni0wZDg5MTBmYi1hNjE5LTQwNTYtYTk3ZS1mN2QzYWRmNTE0ZDEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDExMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAxMTNUMTIxOTE2WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YTEwYjNkOGYwM2Q0YzQ2ZTU4MGJiODFkMzMyMWM4MDNiMjhmMzBiNzYwMTZhOTdkYTg5NDllY2ZiY2VmMzczOSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.DvncrC8SFCzwoX-7QPPORqns1H8gDTCi7LGdMC0JiwQ">

I used longer titles during testing, there it didn't feel that cramped. Since I'm planning to add even more information there I tried to clean it up a bit:
![image](https://github.com/openHPI/codeharbor/assets/24592181/97816753-22d1-4767-8830-2d7b6f4a9eaf)
What do you think?

MrSerth commented:
> What do you think?

Ah, much better! I like that, thanks! 🎨🖌️

Dome-GER commented:
Could you please resolve the conflicts? 😇

Dome-GER commented on app/controllers/collections_controller.rb:
```diff
@@ -3,11 +3,10 @@
 
 class CollectionsController < ApplicationController
-  before_action :load_and_authorize_collection, except: %i[index new create]
+  before_action :load_and_authorize_collection, except: %i[index new create create_ajax]
```
Having a separate action for responding to AJAX is a bit uncommon, I'd say. Usually, one would allow an action to respond to both. An alternative would be to route it to a dedicated controller (e.g., scoped controller in an `Ajax` module). The choice depends on the amount of code / logic to be re-used, of course.

How is this handled in other places in your code, so is this consistent?

Dome-GER commented on app/controllers/collections_controller.rb:
```diff
@@ -3,11 +3,10 @@
   def index
-    @collections = Collection.includes(:users, tasks: %i[user groups])
-      .where(collection_users: {user: current_user})
+    @collections = Collection.member(current_user).or(Collection.public_access).includes(:users, tasks: %i[user groups])
```
It's hard to tell whether this is related to the AJAX action or not. Usually, I try to keep the changes separate, but probably not needed here.

Dome-GER replied:
So this will be extracted or are you planning to keep it here, now?

Dome-GER commented on app/controllers/collections_controller.rb:
```diff
@@ -36,6 +35,19 @@ def create
+    authorize @collection
+
+    if @collection.save
+      redirect_to @collection.tasks.first, notice: t('.success_notice')
```
Is a redirect the expected behavior for an AJAX call? 🤔 I'd expect a JSON response, no?

Dome-GER commented on app/views/collections/_form.html.slim:
```diff
@@ -8,6 +8,14 @@
         = f.label :description, Collection.human_attribute_name('description'), class: 'form-label'
         = f.text_area :description, class: 'form-control', rows: 10
+      .field-element
+        = f.label :visibility_level,  Collection.human_attribute_name("visibility_level"), class: 'form-label collection-edit-visibility'
```
How about a real label here?

Dome-GER replied:
You are using `human_attribute_name` here. This is a helper that makes the attribute more readable, but it's not a real label. You could also just define a new locale as we do for other attributes.

Dome-GER replied:
I just didn't know. So this was a deliberate decision, fine for me.

Dome-GER commented on app/views/collections/show.html.slim:
```diff
@@ -1,10 +1,26 @@
+        - if @collection.users.length == 1
+          =< t('.no_other_user')
+        - else
+          =< t('.num_of_other_users', count: @collection.users.length - 1)
```
Is this a deliberate choice for `length` since we need the collection users anyway?
I was just wondering, since further down you call `count`.

kkoehn replied:
No, I have not thought about the use of `length` here - The `count` further down was already used before. If I see it correctly `size` should fit them both better.

Dome-GER commented on app/views/collections/show.html.slim:
```diff
@@ -49,22 +65,22 @@
+              => t('.button.share')
+              span.caret
+            .dropdown-menu#share-menu
+              .dropdown-header style="padding-left: 2px;"
```
What is your "policy" regarding inline styles? 🤔

kkoehn replied:
We don't really have a policy, but I try to avoid it normally

Dome-GER commented on app/views/tasks/show.html.slim:
```diff
@@ -478,13 +478,21 @@
-                  li
-                    = link_to collection.title, add_to_collection_task_path(collection: collection), method: 'post', class: 'dropdown-item'
-              - else
+              - current_user.collections.reverse.each do |collection|
```
I assume this is using `reverse` since collections have a default order by creation date, and we want to list the most recent first. Could be made explicit (but I saw that this code has been moved, and it was this way before).

kkoehn replied:
Good idea!

Dome-GER commented on app/views/tasks/show.html.slim:
```diff
@@ -478,13 +478,21 @@
+                = form_with model: Collection.new, url: create_ajax_collections_path do |f|
+                  .create-collection.input-group
+                    = f.text_field :title, class: 'form-control', placeholder: t('.create_collection_placeholder')
+                    = f.hidden_field :task_ids, value: @task.id
```
`ids` vs. `id` is confusing for me. But I don't know the domain.

kkoehn replied:
I understand the confusion, but I think it makes sense:
A `Collection` `has_many` `Tasks` and here we want to create a collection with one `Task` already added

Dome-GER commented on config/routes.rb:
```diff
@@ -32,12 +32,17 @@
 
       patch :remove_all
       patch :remove_task
+      patch :remove_task_ajax
```
You can restrict routes as follows, assuming they shall only be available for AJAX calls:

```suggestion
      patch :remove_task_ajax, constraints: ->(r) { r.xhr? }
```

Dome-GER replied:
This is a leftover now, isn't it?

Dome-GER commented on db/schema.rb:
```diff
@@ -82,6 +82,7 @@
     t.datetime "updated_at", null: false
     t.string "title"
     t.string "description", default: "", null: false
+    t.integer "visibility_level", limit: 2, default: 0, null: false, comment: "Used as enum in Rails"
```
In the migration, `limit` is 1, but here it is 2. 🤔 

And why is the schema timestamp not updated?

kkoehn replied:
The changed limit is interesting, setting limit for integers specifies the number of bytes that the number uses. Since Postgres does not have a `tinyint` type it reverts to 2, so it can fit a `smallint`.

the timestamp probably went MIA during the merge

Dome-GER commented on spec/factories/collection_user.rb:
```diff
@@ -0,0 +1,8 @@
+# frozen_string_literal: true
+
+FactoryBot.define do
+  factory :collection_user do
```
Is this a useful factory? I'd rather expect to see the association between a user and collection be created in the test setup explicitly.

kkoehn replied:
I use it in `collection_user_spec.rb` to validate the presence of validations. I think it makes sense and is useful, even if only for that one case.

Dome-GER commented on app/controllers/collections_controller.rb:
```diff
@@ -25,11 +24,13 @@ def new
+    create_with_task = @collection.tasks.any?
 
-    if @collection.save
+    if create_with_task
```
I'd rather add an explanatory code comment here instead of making it a descriptive variable.
Benefit: A code comment can focus on the _WHY_. Why is the redirect behavior different if there are any tasks?

Dome-GER commented on app/views/collections/_form.html.slim:
```diff
@@ -8,6 +8,14 @@
         = f.label :description, Collection.human_attribute_name('description'), class: 'form-label'
         = f.text_area :description, class: 'form-control', rows: 10
+      .field-element
+        = f.label :visibility_level,  Collection.human_attribute_name("visibility_level"), class: 'form-label collection-edit-visibility'
```
```suggestion
        = f.label :visibility_level, Collection.human_attribute_name("visibility_level"), class: 'form-label collection-edit-visibility'
```

Dome-GER reviewed (commented):
Thanks for incorporating the feedback. I didn't look at the specs in detail.

MrSerth commented:
@kkoehn: Would you mind rebasing this PR, please? Sorry for the delay in the review; I am waiting for it, too.

Note to self: Once done, I need to set some collections to public.

MrSerth commented:
The failing specs are related to missing translations (i.e., for German). That should be something easy to fix :)

Dome-GER commented on app/views/tasks/show.html.slim:
```diff
@@ -478,13 +478,21 @@
                 li
-                  = link_to t('.button.add_collection'), new_collection_path(current_user)
+                  - if collection.tasks.include?(@task)
+                    = link_to "<span class=\"collection-checkmark\">✓</span>#{collection.title}".html_safe, remove_task_ajax_collection_path(collection, task: @task, return_to_task: true), method: :patch, class: 'dropdown-item', data: { confirm: t('.remove_task_from_collection_warning') }
```
This still looks a bit sketchy to me.

Probably, I'd go for `link_to ... do`, which is better readable and meant for such use cases.

kkoehn replied:
You are right, this looks way better now

MrSerth replied:
Nice, thanks for taking another look here. Personally, I think the `link_to ... do` solution now implemented is much cleaner and better to maintain 💪

MrSerth commented on config/locales/de/views/collections.yml:
```diff
@@ -29,6 +30,16 @@ de:
       temporarily_disabled: Vorübergehend deaktiviert
       type_hint: Geben Sie die E-Mail-Adresse des/der Benutzer:in ein
+      visibility:
+        private: Diese Aufgabensammlung ist Privat
+        public: Diese Aufgabensammlung ist Öffentlich
```
```suggestion
        private: Diese Aufgabensammlung ist privat
        public: Diese Aufgabensammlung ist öffentlich
```

MrSerth commented on config/locales/de/views/collections.yml:
```diff
@@ -29,6 +30,16 @@ de:
         deletion_warning: Sie sind der/die letzte Benutzer:in in dieser Aufgabensammlung. Wenn Sie gehen, wird diese Aufgabensammlung dauerhaft gelöscht! Sind Sie sicher?
       locked_exercise_hint: 'Hinweis: Für Aufgaben mit einem Schloss-Symbol müssen Sie zuerst Zugriff beantragen, um sie anzeigen zu können.'
       noExercisesAdded: Keine Aufgabe hinzugefügt
+      no_other_user: Kein anderer Benutzer hat Zugriff auf diese Aufgabensammlung
+      num_of_other_users:
+        one: Ein anderer Benutzer hat Zugriff auf diese Aufgabensammlung
+        other: "%{count} andere Benutzer haben Zugriff auf diese Aufgabensammlung"
```
In the German translation, we are using "gendergerechte Sprache" (with `:`), since CodeHarbor is mainly targeted at schools and universities. Please double check those occurrences, i.e., "Benutzer". There are other examples in this file to look at.

MrSerth commented on config/locales/de/views/tasks.yml:
```diff
@@ -98,6 +99,7 @@ de:
       no_license: Keine
       no_model_solution_present: Keine Musterlösungen vorhanden
       no_tests: Keine Tests enthalten
+      remove_task_from_collection_warning: Sind Sie sicher, dass sie diese Aufgabe aus der Sammlung entfernen wollen?
```
```suggestion
      remove_task_from_collection_warning: Sind Sie sicher, dass Sie diese Aufgabe aus der Sammlung entfernen wollen?
```

kkoehn commented on app/views/tasks/show.html.slim:
```diff
@@ -478,13 +478,26 @@
+                    = link_to remove_task_collection_path(collection, task: @task, return_to_task: true), method: :patch, class: 'dropdown-item', data: { confirm: t('.remove_task_from_collection_warning') } do
+                      span.collection-checkmark
+                        = '✓'
+                        = collection.title
```
You are right, there was a space missing
